---
title: "DESeq2 Tutorial"
author: "Yurim Park"
date: "19/5/2022"
output: html_document
---

Useful resources 

- **Vignette** - http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html
- **Book** - http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

# install packages 

- If i didn't already have these installed in my environment - I would delete the # on lines 19-23 and 26-39 and run the code 

```{r}
# from cran 
# install.packages('tidyverse')
# install.packages("RColorBrewer")
# install.packages('pheatmap')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# from bioconductor
# BiocManager::install("ashr")
# BiocManager::install("apeglm")
# BiocManager::install("DESeq2")
# BiocManager::install("org.Mm.eg.db")
```

# load packages

- what do these packages do? Add comments to remind you what they do 
```{r}
library("pheatmap")   
library("RColorBrewer")
library(tidyverse)     
library(DESeq2)      
library(ashr)          
library(apeglm)      
library(org.Mm.eg.db)
```

# 1 Prepping our input data to make deseq2dataset object

## Import counts table

- We made this (or a very similar version) using featurecounts in week1 on the cluster
- raw data -> fastqc -> hisat2 -> featurecounts 

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/allsamples_counts.tsv.gz", comment = "#")


# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  dplyr::select(!c(Chr, Start, End, Strand, Length)) %>%
                  as.matrix()

column_to_rownames(featurecounts_table)

# check the dimensions of this counts_table table using dim() -> should have 39,179 genes and 8 cols 
dim(counts_table)

```

# Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/PRJEB18572_sample_info_subset.tsv")
sample_info
str(sample_info)
sample_info$cell_type<- as.factor(sample_info$cell_type)
# What class is the cell_type currently? 
# We want to change it to a factor with 2 levels (the hypothesis we will test)
# Which should be our reference level? 
# what other columns might we want to include in our design that we want to change to factors? 
sample_info$genotype <- as.factor(sample_info$genotype) # as.factor makes factor alphabetically 
sample_info$replicate <- as.factor(sample_info$replicate) # as.factor makes factor alphabetically



```

# Check the sample names are the same and in the same order in the sample_info and counts_table 

```{r}
# Column names of the count table and row names of the sample info table must be in the same order
# Check this 
colnames(counts_table) == rownames(sample_info)
colnames(counts_table)
colnames(sample_info)

# make sample_info row names from sample_id 
sample_info <- sample_info %>% 
               column_to_rownames("sample_id") %>% 
               as.data.frame()
# remove .bam from colnames 
colnames(counts_table) <- counts_table %>%
    colnames() %>%
    str_replace(".bam","")

# another solution
#colnames(counts_table) <- gsub(".bam","",colnames(counts_table))


```

# Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
dds <- DESeqDataSetFromMatrix(countData = counts_table, colData = sample_info, design = ~ genotype + replicate + cell_type) 
  
```

## Investigate the dds object 

```{r}
# what does it look like? - e.g. use str(dds)
str(dds)

# how to access counts
counts(dds)

# access colData i.e the sample info table
colData(dds)

# access rowData
rowData(dds)

# access the design
design(dds)

# does it have any rowRanges()?
rowRanges(dds)


# what are its rownames and colnames?
rownames(dds)
colnames(dds)

# what attributes does it have?
attributes(dds)

```

## How to subset a DESeqDataSet Object 

```{r}
# practice subsetting dds object
# select first 3 columns
dds_subset <- dds[,1:3]

```

# Filter out unexpressed genes 

- It's good practice to filter out genes with 0 counts across all samples at the start 
- This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis 
- Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.


```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`
dds_filtered <- dds[rowSums(counts(dds))>0, ]

# can also be more stringent in filtering depending on the dataset, filter out low expressing genes with less than 10 counts across 3 samples or higher for example
# at least 3 samples with a count of 10 or higher
#keep <- rowSums(counts(dds) >= 10) >= 3
#dds <- dds[keep,]
#nrow(dds) 

# check dims of dds_filtered  
dim(dds_filtered)  
  
```

# Perform exploratory analysis on dds_filtered object 

1. transform dds_filtered object using `rlog()` functions
    - name your new variables `rlog_transformed_dds` 
    - we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    - check the class of the rlog_transformed_dds objects 
    - you can look at the tranformed counts table using `assay()` 

NOTE we only use this tranformed data for QC not for downstream analysis or differential expression itself!!


```{r}
#transform data using `rlog()` functions
rlog_transformed_dds <- rlog(dds_filtered, blind = TRUE)


# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 
class(rlog_transformed_dds)
  
#get and look at the tranformed counts table using `assay()`
assay(rlog_transformed_dds) %>% head()

```

# Optional: If you have time do the tranformation again but using the vst() function 

- As a guide 
    - rlog - more robust than VST when size factors vary widely (10x)
    - VST - less sensitive to high count outliers than rlog, quicker - recommended most often 

```{r}
# get vst transformed data using vst() 
vst_transformed_data <- vst(dds_filtered, blind=TRUE)

```


# Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()


```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 
corr_dds <- cor(assay(rlog_transformed_dds), method="spearman")
corr_dds_pearson <- cor(assay(rlog_transformed_dds), method="pearson")

# use pheatmap to plot the correlation matrix 
pheatmap(corr_dds)
pheatmap(corr_dds_pearson)
```

# try ploting a heatmap of sample distances 

```{r}
# get sample distances -> Annotated what this code is doing?
sampleDists <- dist(t(assay(rlog_transformed_dds))) 
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)
```

What do you notice here!? what does it tell us about our data? 

Answer: 083/089 and 082/088 are duplicates of one another (their correlation is 1); one in each pair can be removed



## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6]

# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above
sampleDists <- dist(t(assay(rlog_transformed_dds_subset))) 
sample_distances_matrix <- as.matrix(sampleDists) 

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = colors)

```

What do you notice now about the data and the clustering?

Answer: Samples are now clustering by celltype. 82/83/84 are CD4 and 85/86/87 are CD8

# Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r, fig.asp=1}
# plot pca of rlog_transformed_dds using plotPCA()
plotPCA(rlog_transformed_dds_subset, intgroup='cell_type')



# what does the intgroup parameter do? have a go at changing it...
plotPCA(rlog_transformed_dds_subset, intgroup='cell_type')
plotPCA(rlog_transformed_dds_subset, intgroup='genotype')
plotPCA(rlog_transformed_dds_subset, intgroup='replicate')

# how many variable genes is it using for pca? 
# 500

# plot the pca for all genes 
plotPCA(rlog_transformed_dds_subset, intgroup='sample_name',ntop = nrow(rlog_transformed_dds_subset))
# Selects top genes based on their variance and then uses prcomp with default settings (centering but not scaling) for PCA


# plot pca of rlog_transformed_dds using plotPCA()
#plotPCA(rlog_transformed_dds_subset,ntop=nrows(rlog_transformed_dds_subset),intgroup="cell_type")+coord_fixed(ratio=5)
# what does the intgroup parameter do? have a go at changing it...
#plotPCA(rlog_transformed_dds_subset,ntop=500,intgroup="genotype")+coord_fixed(ratio=5)
#plotPCA(rlog_transformed_dds_subset,ntop=500,intgroup="replicate")+coord_fixed(ratio=5)
#head(rlog_transformed_dds_subset)
# how many variable genes is it using for pca? 
# plot the pca for all genes 

```


What does it tell us about our sample?

Answer: 


------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis 
    e.g. - update design with variables in sample info table we want to regress out 
         - remove samples (poor quaility or those with name switches etc)
         
# Revisiting our inital data 

What do we need to do to our data??? 

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}

# do we need to remove/edit samples? 

# do we want to update the design? Is this nessecary in this case?
# e.g. design(dds) <- ~replicate + condition
dds_filtered_updated <- dds_filtered[,1:6]
dds_filtered_updated
design(dds_filtered_updated)<- ~replicate+cell_type
dds_filtered_updated
```


# Now we can start our actual DE!!!! yay 

Run deseq2 modelling using one function (DESeq())

```{r}

# run deseq2 modelling on the updated dds object in one function using DEseq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

# then extract the results
results_obj <- results(dds_filtered_updated)

# sumarise results -> results depend on model used!
summary(results_obj)

# for model ~cell_type 
  # LFC > 0 (up) : 4414, 20%. 
  # LFC < 0 (down)     : 4490, 20%

# for model ~replicate + cell_type
  # LFC > 0 (up)       : 3859, 17%
  # LFC < 0 (down)     : 3890, 17%






# look at dds slots then run deseq2 function 
# what has happened to your dds slots?

# run deseq2 modelling on the updated dds object in one function using DEseq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

# then extract the results
results_obj <- results(dds_filtered_updated, alpha = 0.05)
results_obj_2 <- results(dds_filtered_updated, alpha = 0.05, contrast=c("cell_type", "CD4", "CD8"))

# sumarise results 
summary(results_obj)
saveRDS(results_obj, file = "result_obj_DESeq.RDS")
saveRDS(dds_filtered_updated, file = "dds_filter_update_DESeq2_file.RDS")

```

Have a look at the outputs 
```{r}
resultsNames(dds_filtered_updated)
```

# Now we will run the analysis step by step to see whats going on 

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()
dds_filtered_updated <- estimateSizeFactors(dds_filtered_updated)

# Access the size factors from the dds_filtered_updated object
sizeFactors(dds_filtered_updated)

# Do you see any systematic differences in size factors between groups?


```

# Optional: Plot size factors as barchart using ggplot
```{r}
size_factors <- sizeFactors(dds_filtered_updated)
size_factors <- data.frame(sample = names(size_factors),
                           size_factor = size_factors,
                           sample_group =colData(dds_filtered_updated)$cell_type)

ggplot(size_factors, aes(x = sample, y = size_factor, fill = sample_group)) +
    geom_col() +
    theme(axis.text.x = element_text(angle = 45, h = 1))
```

## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts
DESeq2::counts(dds_filtered_updated)
normalized_counts <- DESeq2::counts(object = dds_filtered_updated, normalized = TRUE)

# check the class of the output
class(normalized_counts)
```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions() 
Dispersion_obj <- estimateDispersions(dds_filtered_updated)
# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)
plotDispEsts(Dispersion_obj)
```
```{r}
# Access dispersions from the DESeqDataSet object for individual genes
head(dispersions(dds_filtered_updated))
rowData(dds_filtered_updated)    # mcols(dds) accesses the same slot
View(as.data.frame(rowData(dds_filtered_updated)))

# Description of columns in mcols
mcols(mcols(dds_filtered_updated), use.names = TRUE)


# dimensions of dispersiosn = one dispersion per gene
# NA values where 0 counts for all samples for that gene
```


## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest
dds_waldtest <- nbinomWaldTest(dds_filtered_updated)
results(dds_waldtest)
mcols(dds_waldtest)$description
summary(results(dds_waldtest))

##### EXTRA ####### 
# if have more than one group in contrast can pull out those comparisions by specifying contrasts 
# check contrast names 
resultsNames(dds_waldtest) # to see what names to use
contrast <- list(resultsNames(dds_waldtest)[1], resultsNames(dds_waldtest)[2])
contrast

```
Contrasts can be supplied as a character vector with exactly three elements: 
  - the name of the factor (of interest) in the design formula, 
  - the name of the two factors levels to compare. 
  - The factor level given last is the base level for the comparison.
  
	contrast <- c("condition", "level_to_compare", "base_level")
	results(dds, contrast = contrast)

# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}
# filter rows with 0 counts 
dds_actual_analysis <- dds[rowSums(counts(dds)) > 0, ]

# remove unwanted samples 
dds_actual_analysis <- dds_actual_analysis[,1:6]

# make DESeqDataSet (dds) object 
design(dds_actual_analysis) <- ~ replicate + cell_type

# run DESeq() 
dds_actual_analysis <- DESeq(dds_actual_analysis)
```

# Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 
results_obj_actual <- results(dds_actual_analysis, alpha=0.05)
summary(results_obj_actual)

results_obj <- readRDS("result_obj_DESeq.RDS")
dds_filtered_updated <- readRDS("dds_filter_update_DESeq2_file.RDS")

```

# Optional - if we have time: Access the coefficients of the Negative Binomnial GLM – see vingette("DESeq2")
```{r}
head(coef(dds_actual_analysis))
coef_dds <- coef(dds_actual_analysis)

# Intercept is absolute gene expression in baseline group in rea
# Coefficients are the logFC in the other conditions
# Coefficients are the log2 fold changes for each group

##### EXTRA #############################

# Convert coefficients to "real" space rather than log values
coef_real <- 2^(coef(dds_actual_analysis))
coef_real["ENSMUSG00000041779", ]
# Intercept     376.604199
# cell_type_CD8_vs_CD4    1.251456

# Intercept - log2 normalised expression level of gene for baseline condition, which is Egr2_3_DKO_CD4 (uses size factors and dispersion estimates calculated earlier)
```

# Plot normalised counts for ENSMUSG00000041779
 
 checkout the 'plotCounts()` function
 
```{r}
# Plot normalised counts for gene ENSMUSG00000041779
plotCounts_matrix <- plotCounts(dds = dds_filtered_updated, gene = "ENSMUSG00000041779", intgroup = "cell_type", normalized = TRUE, returnData = TRUE)

plotCounts_matrix

```

# Access the results table for the comparison between CD8 and CD4 T cells . Youu can also try to CD4 vs CD8 (i.e. switch the reference) - you could also try this for replicate 1 vs replicate 3 as a control 

checkout `resultsNames()`  & `results()`

```{r}
resultsNames(dds_filtered_updated)
head(results(dds_filtered_updated))
results_obj <- results(dds_filtered_updated)

result_obj_replicate3v1 <- results(dds_filtered_updated, name = "replicate_3_vs_1")
dim(result_obj_replicate3v1)


result_obj_celltype_CD4_vs_CD8 <- results(dds_filtered_updated, contrast = c("cell_type", "CD4", "CD8"))

summary(results(dds_filtered_updated, alpha = 0.05))
```

Table should contain log2 fold changes and p-values for all genes

# Plot a histogram of the raw and BH-adjusted p-values using ggplot – do they look as expected? for CD8 vs CD4

```{r}
results_obj_dataframe <- as.data.frame(results_obj)
head(results_obj_dataframe)

ggplot(as.data.frame(results_obj))+ geom_histogram(aes(x=pvalue), fill = "pink")+geom_histogram(aes(x=padj), fill = "light blue")

```


# Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}
plotMA(results_obj, alpha = 0.05)
```

# Shrink the log2 FC values using the normal, apeglm and ashr methods -> see the vignette – which shall we choose? 

```{r}

# call your results table  called shrunkenLFC_dds_results_obj
shrunken_results_obj <- lfcShrink(dds_filtered_updated, coef = 4,  type="apeglm")

shrunken_results_obj
plotMA(shrunken_results_obj, alpha = 0.05)

```


# Generate a results table (one shrinkage method) containing gene symbols and (if time) entrez ids 

```{r}

# convert shrunkenLFC_dds_results_obj to a dataframe and call it shrunkenLFC_dds_results_df
shrunken_results_obj_df <- as.data.frame(shrunken_results_obj)

# Use the org.Mm.eg.db package to get gene symbols for the ensembl gene ids (hint:you did this on bioconductor day) 

# How many Ensembl IDs are not assigned an gene symbol and how many gene symbols are duplicated?


library(org.Mm.eg.db)
keytypes(org.Mm.eg.db)      #to remind us of what keytypes are in this particular db

# use mapIds() to get 1:1 mapping for SYMBOL, ENTREZ and MGI
gene_symbols <- mapIds(org.Mm.eg.db, keys = row.names(shrunken_results_obj), "SYMBOL", "ENSEMBL" )
gene_entrezid <- mapIds(org.Mm.eg.db, keys = row.names(shrunken_results_obj), "ENTREZID", "ENSEMBL" )
gene_mgi <- mapIds(org.Mm.eg.db, keys = row.names(shrunken_results_obj), "MGI", "ENSEMBL" )

# add mapId vectors to  shrunkenLFC_dds_results_df
shrunken_results_obj_df <- cbind(shrunken_results_obj_df, gene_symbols = gene_symbols, gene_entrezid = gene_entrezid, gene_mgi = gene_mgi)


dim(shrunken_results_obj_df)
dim(shrunken_results_obj)
##shrunken_results_obj_df2 <- shrunken_results_obj_df %>% mutate(Symbol = gene_symbols, Entrezid = gene_entrezid, MGI = gene_mgi)

```


# Export all the results to CSV

```{r}
write_csv(shrunken_results_obj_df, "CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv")
```

# Filter res_df and save as CSV

```{r}
#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- shrunken_results_obj_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- shrunken_results_obj_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- shrunken_results_obj_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```


Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account. 

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....
